[ft-rbc23] More gk area points, shape refactor in structs

Wanchai290 opened the pull request:
Instead of having points in the struct, the structures `Penalty` and `Goal` use instead the struct `Rectangle` to define these points.
Additionally, the original attributes have been kept so that the viewer [(aquarium)](https://github.com/NAMeC-SSL/aquarium) can still display properly what it receives.

This PR requires some additional changes on aquarium, which have been addressed by a new branch on its repository [(crabe-refac-shapes)](https://github.com/NAMeC-SSL/aquarium/tree/crabe-refac-shapes).

there are some weird twists in the code, notably because of notation issues. looks messy but I tried my best, looking for advice on how to improve this

Jossl123 committed:
add goal geometries

Jossl123 committed:
add geometry for goal

Wanchai290 committed:
refactor!: use Rectangle struct for Goal/Penalty

This breaks the integration with the viewer. Requires re-defining the
types in the viewer

Wanchai290 committed:
fix: serde: rename all fields with camelCase

EtienneSchmitz commented on crates/crabe_framework/src/data/geometry.rs:
```diff
@@ -46,30 +46,47 @@ impl Default for Geometry {
-            },
+
+            // You might ask "Why is goal width not the rectangle's width here ???"
+            // To this I answer "No fucking idea"
```
I understand there may be frustration, but it's important for us to maintain a professional tone in the comments and stay focused on the project objectives. 
This comments isn't constructive or relevant to our goals.
Let's aim to keep our feedback helpful and directly related to the task at hand.

EtienneSchmitz replied:
In response to your question about goal_length and goal_width, it's crucial to note that goal_length is not a standard term in football lexicon, which aligns with the protobuf structure. Our context diverges between the physical dimensions of a football game and the 2D representation in a rectangle.

Considering this, I propose that we adhere to the conventional library approach of using width and height for rectangles. Additionally, to accommodate our specific needs, we can introduce functions like depth() and width() for the goal. This approach allows us to maintain a distinction between the 'width' of the goal in a football context and the 'width' of our 2D rectangle, ensuring both clarity and accuracy in our representations

Wanchai290 replied:
alright i'll change this to something more pro

i didn't quite catch what you said in the second post though, what I mentioned is already what I implemented. Should I change something ?

Wanchai290 replied:
as we discussed, this is what we wanted already, so I will edit this PR to change the struct accordingly

EtienneSchmitz commented on crates/crabe_math/src/shape/rectangle.rs:
```diff
@@ -1,16 +1,49 @@
     pub height: f64,
     /// The position of the rectangle's top-left corner.
-    pub position: Point2<f64>,
+    pub top_left: Point2<f64>,
```
Based on @benjaminchew4's suggestion, I recommend utilizing memoization to enhance performance. A convenient way to implement this is by using the memoize crate available at [this link](https://crates.io/crates/memoize).

Wanchai290 replied:
![image](https://github.com/NAMeC-SSL/CRAbE/assets/24627625/60d76c0e-015f-439b-8ce4-a5e122b16767)
I can't use this crate for methods. Technically, I could just create a function `make_rectangle` inside `rectangle.rs` and add the memoize attribute to it, but I don't think it looks like very clean code.

Wanchai290 replied:
I mean I think we can take care of this in particular in another PR instead of this one. Looks like too much responsibility on a single PR

benjaminchew4 replied:
The overhead of memoization would likely outweigh the benefits, as it involves a HashMap lookup. However, storing all the points of a shape decreases readability and increases the likelihood of bugs. For rectangles, I think it would be better to have a struct containing the origin, the width and the height. Points could be accessed with methods, e.g:
```rust
pub fn top_right(&self) -> Point2 {
  Point2::new(self.origin.x + self.width, self.origin.y)
}
```

Wanchai290 replied:
this has been taken care of by [ad7f3c7](https://github.com/NAMeC-SSL/CRAbE/pull/93/commits/ad7f3c72b001c426f8a1df1d4cd91ffedc625cbc)

EtienneSchmitz reviewed (changes requested):
Thank you for submitting your pull request.
I've reviewed the code and provided several suggestions and potential improvements.

Wanchai290 committed:
style: change raging comment to profesionnal one

Wanchai290 committed:
refactor: goal width and depth from functions, rewritten Goal::new func

Wanchai290 committed:
refactor: impl Serialize trait to retain comm protocol structure in aquarium viewer

Wanchai290 committed:
feat: change some attributes to methods in rectangle

EtienneSchmitz committed:
chore : remove comment that is not necessary and doesn't follow the guidelines of the project.
Comments doesn't need to give the reason of an implementation, but explain what was build and how to use it.
